### PR TITLE
feat(evm): collaborative indexing through txid sharing

### PR DESCRIPTION
**Pull request by daniel-savu**
### Description

- improved indexing reliability by sharing channels between indexing tasks, such that they inform one another when they come across a hyperlane-related transaction
- the receiving end will then get the receipt of that tx and filter it for events relevant to it
- this adds some flexibility to indexing since we can stop relying on watermark cursor entirely: as long as we have one task using sequence indexing, all others can rely on being informed by their channels about what txs to index
- this PR is only for the EVM chains, but can be easily extended

### Drive-by changes

- Enforces gas payments for all chains in e2e (previously this only used to be the case for sealevel)
- The channel size is configured per type, similarly to the indexing cursor, in `contract_sync/cursors/mod.rs`
- Reduces log verbosity is some places, e.g. by manually implementing `Debug` to exclude fields that aren't relevant to debugging (maybe there's even a macro annotation to exclude, but I haven't checked)
- Renames `MpmcReceiver` -> `BroadcastReceiver`

### Related issues

- Fixes https://github.com/hyperlane-xyz/hyperlane-monorepo/issues/3267

### Testing

e2e tested with the watermark cursor manually disabled

### CODE REVIEW

File: rust/hyperlane-base/src/contract_sync/mod.rs
Comment (tkporter) [inline rust/hyperlane-base/src/contract_sync/mod.rs:261]: nit typo `inedxing`

File: rust/hyperlane-base/src/contract_sync/cursors/mod.rs
Comment (tkporter) [inline rust/hyperlane-base/src/contract_sync/cursors/mod.rs:22]: nit: may be nice to `///` document these functions

File: rust/hyperlane-base/src/contract_sync/mod.rs
Comment (tkporter) [inline rust/hyperlane-base/src/contract_sync/mod.rs:70]: nit: is this accurate? made me think that a new broadcast channel was created altogether, but i guess it's just a new receiver end - led me to misunderstand how it works for a few mins

File: rust/agents/relayer/src/relayer.rs
Comment (tkporter) [inline rust/agents/relayer/src/relayer.rs:328]: nit: wonder if we should just panic or something to make sure this never happens? as this is a startup invariant I assume we'll wanna keep. Don't feel strongly here

File: rust/agents/relayer/src/relayer.rs
Comment (tkporter) [reply rust/agents/relayer/src/relayer.rs:427]: thinking about it more, because we clone the entire hashmap for each origin chain, iiuc we end up creating n^2 broadcast receivers (instead of `n`, where n = the number of origin chains) when running all the IGP syncs. Because they don't get cleaned up (`rxs` is moved into the task), the receivers for all the other origins won't ever read from the channel, so the data will stick around in memory https://docs.rs/tokio/latest/tokio/sync/broadcast/index.html#lagging. So we should def change this to just the broadcast receiver

We could also maybe just make it a non-optional receiver too if we really wanna make sure that we always have a receiver here?

File: rust/agents/relayer/src/relayer.rs
Comment (tkporter) [inline rust/agents/relayer/src/relayer.rs:320]: thinking a little about how the broadcast receiver works, I'm realizing that the way we use `MpmcChannel` for the relayer server may mean that we don't clean up the MessageRetryRequests even when they're served (until we have `ENDPOINT_MESSAGES_QUEUE_SIZE = 1_000` of them). This is because we only ever use `receiver` as something to clone here https://github.com/hyperlane-xyz/hyperlane-monorepo/blob/main/rust/hyperlane-core/src/types/channel.rs#L31. So when we call `run_destination_submitter` on line 344 for each destination chain, we end up creating `n` new receivers, for a total of n+1 receivers. We read from the new `n` receivers in the destination tasks but never the original reciever that lives in the MpmcChannel

Imo this low prio & urgency to fix and doesn't need to be in this PR - maybe something good in a small separate PR or we can put it in the backlog? Will defer to you whether you think it's worth doing rn

File: rust/agents/scraper/src/agent.rs
Comment (tkporter) [inline rust/agents/scraper/src/agent.rs:198]: I know we didn't chat about this but imo it makes sense to also move the scraper and validator to using hook indexing, wdyt?

**Review by tkporter — COMMENTED** [review]
Very nice

File: rust/agents/relayer/src/relayer.rs
Comment (daniel-savu) [reply rust/agents/relayer/src/relayer.rs:427]: Great point, changed to passing a single receiver. The receiver is intentionally non-optional, since e.g. the message sync won't be receving anything, but just broadcasting

File: rust/chains/hyperlane-ethereum/src/contracts/interchain_gas.rs
Comment (daniel-savu) [reply rust/chains/hyperlane-ethereum/src/contracts/interchain_gas.rs:155]: I assume so, `decode_log` is defined on `EthEvent`, which is derived for `GasPaymentFilter` in the abi interface file with this annotation:
```
    #[ethevent(
        name = "GasPayment",
        abi = "GasPayment(bytes32,uint32,uint256,uint256)"
    )]
```

File: rust/hyperlane-base/src/contract_sync/cursors/sequence_aware/backward.rs
Comment (daniel-savu) [reply rust/hyperlane-base/src/contract_sync/cursors/sequence_aware/backward.rs:35]: verbosity is the reason yep. I think this makes a good case for parsing agent logs in e2e -> if we really need something for debugging purposes, we'll try parsing it and realise it's not there

File: rust/chains/hyperlane-ethereum/src/contracts/merkle_tree_hook.rs
Comment (daniel-savu) [reply rust/chains/hyperlane-ethereum/src/contracts/merkle_tree_hook.rs:150]: done

File: rust/agents/relayer/src/relayer.rs
Comment (daniel-savu) [reply rust/agents/relayer/src/relayer.rs:328]: unwrapping only makes sense if we want to enforce hook indexing. My assumption in this PR is that it's optional and we can alternate between this and cursors. If we stop relying on watermark cursor, can change this to a panic

File: rust/agents/relayer/src/relayer.rs
Comment (daniel-savu) [reply rust/agents/relayer/src/relayer.rs]: fixed, that's a great point and explains why there isn't an `mpmc` channel in tokio

File: rust/agents/scraper/src/agent.rs
Comment (daniel-savu) [reply rust/agents/scraper/src/agent.rs]: we can't enable hook indexing for the validator bc we don't have a message indexer, but as discussed it's probably fine without because the checkpoint submitter already uses sequences.

I did enable it for the scraper 👍

File: rust/hyperlane-base/src/contract_sync/mod.rs
Comment (tkporter) [reply rust/hyperlane-base/src/contract_sync/mod.rs]: I definitely didn't realize tx meant transmitter 😅
Bc we use tx in many places to mean transaction might be nice to keep tx for transaction in the future

File: rust/utils/run-locally/src/main.rs
Comment (tkporter) [inline rust/utils/run-locally/src/main.rs:406]: after https://github.com/hyperlane-xyz/hyperlane-monorepo/pull/3870 is merged, it may be nice to follow up here to make sure that we hit the hook indexing code path

**Review by tkporter — APPROVED** [review]
Approving but I believe https://github.com/hyperlane-xyz/hyperlane-monorepo/pull/3833#discussion_r1630942387 may need to be changed before merging

File: rust/hyperlane-base/src/contract_sync/mod.rs
Comment (daniel-savu) [reply rust/hyperlane-base/src/contract_sync/mod.rs]: changed, you're right

File: rust/agents/relayer/src/relayer.rs
Comment (daniel-savu) [reply rust/agents/relayer/src/relayer.rs]: renamed to just `sender`, the type itself should be informing enough

File: rust/agents/relayer/src/relayer.rs
Comment (daniel-savu) [reply rust/agents/relayer/src/relayer.rs]: my original reasoning was to allow `clone`ing receivers if needed, but we need a single receiver so I changed this to be a `Receiver`

File: rust/utils/run-locally/src/main.rs
Comment (daniel-savu) [reply rust/utils/run-locally/src/main.rs:406]: will do, but I did test manually with the watermark cursor disabled and it worked

**Pull request merged**
